### DISCUSSION TITLE
[Bug] Skywalking-java-agent SWDescriptionStrategy causes memory leaks

### DISCUSSION THREAD

**Discussion by hyhyf** [apache/skywalking#12627] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

I used the nashorn script engine for stress testing in my project. After running continuously for 2 hours, SWTypeDescriptionWrapper occupied more than 90% of the memory, and the old GC was frequently.

### What you expected to happen

When dynamic class creation methods are used, they are put into the type cache. If this operation is repeated indefinitely, it will form a large collection that cannot be garbage collected, thereby consuming memory resources.

### How to reproduce

Jdk: 1.8
Java agent: 9.2.0
Excute scripts using nashorn script engine

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment, marked as answer]
We don't support unlimited dynamic classes. I think this is the current case. And I don't see any reason to do this in product env.

**Reply by wu-sheng** [reply]
The cache is necessary to speed up the analysis process and booting, also to better support for retransfer.

**Reply by hyhyf** [reply]
> We don't support unlimited dynamic classes. I think this is the current case. And I don't see any reason to do this in product env.

This is a business interface we use in the production env to help execute scripts and analyze results. With a high volume of requests, an unexpected incident occurred just two days after the agent was upgraded.

**Reply by hyhyf** [reply]
> The cache is necessary to speed up the analysis process and booting, also to better support for retransfer.

Yes, I understand, but is it possible to provide some mechanism to prevent such accidents, such as discarding expired cache after exceeding size limits? If a microservice has the capability to dynamically create classes, even with reasonable use, the type cache may eventually occupy more than 50% of the memory over time, which seems unreasonable.

**Reply by wu-sheng** [reply]
I don't think so. Because this would cause expected changes to the codes and execution processes.
Such as half of plugin interceptors are instrumented due of this mechanism, then still memory leak happens.

**Reply by hyhyf** [reply]
You're right, but I think someone will face the same problem as me in the future.

**Reply by wu-sheng** [reply]
Yes, maybe, but there is no proper solution for that.

**Reply by gaojingyuan001** [reply]
I’ve encountered the same issue. In my case, the script execution creates different class loaders based on the business logic to load and run scripts. When the number of custom class loaders increases, it causes a memory leak in SWDescriptionStrategy.

**Reply by kylixs** [reply]
I will work on this issue.

**Reply by kylixs** [reply]
@gaojingyuan001 @hyhyf Could you please help test this PR https://github.com/apache/skywalking-java/pull/772 ? Looking forward to hearing from you.
